File: trade Journal 2.0/tradezella-clone/dashboard/lib/normalizer.ts
/**
 * lib/normalizer.ts — TypeScript port of mac/normalizer.py
 *
 * PURPOSE: Convert the raw array of MT5 deals (from VPS GET /history) into
 * clean "TradeRow" objects that map 1:1 to rows in the `trades` Supabase table.
 *
 * WHY THIS IS NEEDED — MT5 DEAL STRUCTURE:
 *   MT5 represents a complete round-trip trade as TWO separate "deal" records:
 *
 *     deal.entry == 0  (DEAL_ENTRY_IN)  — position was opened at this price/time
 *     deal.entry == 1  (DEAL_ENTRY_OUT) — position was closed at this price/time
 *
 *   Both deals share the same `position_id`. The normalizer groups deals by
 *   position_id and merges the IN + OUT pair into a single TradeRow with:
 *     - open_price / open_time from the IN deal
 *     - close_price / close_time / profit from the OUT deal
 *     - commission and swap summed across both deals
 *
 * PARTIAL CLOSES:
 *   When a position is partially closed (e.g. 50% of the position at a time),
 *   MT5 generates multiple OUT deals for the same position_id. Currently we
 *   keep only the LAST OUT deal per position. This means partial close data
 *   is aggregated to a single close event — acceptable for now, can be refined later.
 *
 * USAGE:
 *   const rows = normalizeDeals(historyResponse.deals);
 *   // rows is now ready for Supabase upsert via tradeRowToSupabase()
 */

// ---------------------------------------------------------------------------
// Pip size table — maps symbol → pip value for pip P&L calculation
// ---------------------------------------------------------------------------

/**
 * Pip sizes for common symbols.
 * For forex: 1 pip = 0.0001 for 4-decimal pairs, 0.01 for JPY pairs.
 * For metals: XAUUSD 1 pip = $0.10 (price in USD, tick = 0.01).
 * For indices: 1 pip = 1 full point on US30, 0.1 on US500/NAS100.
 * For crypto: 1 pip = $1 on BTCUSD.
 *
 * Unknown symbols fall back to DEFAULT_PIP_SIZE (0.0001) — pips will be
 * inaccurate but the dollar P&L values are always correct.
 */
const PIP_SIZES: Record<string, number> = {
  // Forex majors / minors
  EURUSD: 0.0001, GBPUSD: 0.0001, AUDUSD: 0.0001,
  NZDUSD: 0.0001, USDCAD: 0.0001, USDCHF: 0.0001,
  EURGBP: 0.0001, EURJPY: 0.01,   GBPJPY: 0.01,
  USDJPY: 0.01,   AUDJPY: 0.01,   CADJPY: 0.01,
  CHFJPY: 0.01,   NZDJPY: 0.01,
  // Metals (spot)
  XAUUSD: 0.1,    XAGUSD: 0.001,
  // Equity indices (CFDs)
  US30: 1.0,      US500: 0.1,     NAS100: 0.1,
  GER40: 0.1,     UK100: 0.1,
  // Crypto
  BTCUSD: 1.0,    ETHUSD: 0.1,
};

/** Fallback for symbols not in the table above */
const DEFAULT_PIP_SIZE = 0.0001;

/**
 * Looks up the pip size for a symbol.
 * Strips trailing "M" (micro lots suffix some brokers append) and trailing
 * dots before looking up. Falls back to DEFAULT_PIP_SIZE if unknown.
 *
 * @param symbol — MT5 symbol string e.g. "EURUSD", "EURUSDM", "XAUUSD."
 */
function getPipSize(symbol: string): number {
  // Remove broker-specific suffixes: trailing "M" (micro) or "." (some brokers)
  const clean = symbol.toUpperCase().replace(/M$/, "").replace(/\.$/, "");
  return PIP_SIZES[clean] ?? PIP_SIZES[symbol.toUpperCase()] ?? DEFAULT_PIP_SIZE;
}

// ---------------------------------------------------------------------------
// Types — input (RawDeal) and output (TradeRow) shapes
// ---------------------------------------------------------------------------

/**
 * Raw deal object as received from the VPS /history endpoint.
 * Fields mirror the MT5 TradeDeal structure with some optional fields
 * because not all fields are available for all deal types.
 */
export interface RawDeal {
  ticket: number;       // unique deal ticket
  order?: number;       // order that generated this deal
  position_id: number;  // links entry and exit deals for the same position
  symbol: string;       // instrument (e.g. "EURUSD")
  type?: number;        // deal type: 0 = buy order, 1 = sell order
  entry: number;        // 0 = ENTRY_IN (opening), 1 = ENTRY_OUT (closing)
  lot_size?: number;    // volume in lots
  price?: number;       // execution price of this deal
  time?: string;        // ISO-8601 UTC timestamp
  commission?: number;  // broker commission for this half of the trade (negative)
  swap?: number;        // overnight financing charge (positive or negative)
  profit?: number;      // profit from this deal (0 for entry deals)
  sl?: number;          // stop-loss level (only set on entry deals; 0 = not set)
  tp?: number;          // take-profit level (only set on entry deals; 0 = not set)
  comment?: string;     // trade comment
  magic?: number;       // EA identifier (0 = manual trade)
}

/**
 * One completed trade ready for storage in Supabase.
 * Produced by merging the IN + OUT deals for a single position_id.
 */
export interface TradeRow {
  position_id:      number;           // MT5 position ID (unique per account)
  ticket:           number | null;    // ticket of the closing (OUT) deal
  symbol:           string;
  direction:        "buy" | "sell";   // the original trade direction
  lot_size:         number;
  open_price:       number | null;    // price at which position was entered
  close_price:      number | null;    // price at which position was closed
  sl:               number | null;    // stop-loss (from the entry deal)
  tp:               number | null;    // take-profit (from the entry deal)
  open_time:        string | null;    // ISO-8601 timestamp of entry
  close_time:       string | null;    // ISO-8601 timestamp of exit
  duration_minutes: number | null;    // time held in whole minutes
  pnl:              number;           // gross profit/loss (from MT5, before fees)
  pnl_pips:         number | null;    // pnl expressed in pips (for position sizing analysis)
  commission:       number;           // total commission (entry + exit deals summed)
  swap:             number;           // total swap (entry + exit deals summed)
  net_pnl:          number;           // pnl + commission + swap = actual money made/lost
}

// ---------------------------------------------------------------------------
// Constants and helpers
// ---------------------------------------------------------------------------

const ENTRY_IN  = 0; // MT5 DEAL_ENTRY_IN  — deal opens the position
const ENTRY_OUT = 1; // MT5 DEAL_ENTRY_OUT — deal closes the position

/**
 * Computes how many whole minutes elapsed between two ISO-8601 timestamps.
 * Returns 0 if the result would be negative (data quality issue).
 */
function durationMinutes(openIso: string, closeIso: string): number {
  const delta = new Date(closeIso).getTime() - new Date(openIso).getTime();
  return Math.max(0, Math.floor(delta / 60_000));
}

/**
 * Converts a price move into pips, adjusting for trade direction.
 *
 * For a BUY trade: pips = (close - open) / pipSize   (positive = profit)
 * For a SELL trade: pips = (open - close) / pipSize  (positive = profit)
 *
 * Rounded to 1 decimal place to avoid floating-point noise.
 */
function calcPips(
  openPrice: number,
  closePrice: number,
  direction: "buy" | "sell",
  symbol: string,
): number {
  const pip = getPipSize(symbol);
  const raw = (closePrice - openPrice) / pip;
  // Negate for sells: a lower close price is profit when short
  return Math.round((direction === "buy" ? raw : -raw) * 10) / 10;
}

// ---------------------------------------------------------------------------
// Core normalizer — groups deals by position_id and merges pairs
// ---------------------------------------------------------------------------

/**
 * Converts an array of raw MT5 deals into an array of complete TradeRows.
 *
 * Algorithm:
 *  1. Group all deals by position_id into { in: RawDeal | null, out: RawDeal | null }
 *  2. Skip positions that have no OUT deal (still open)
 *  3. For each paired position, derive direction from the exit deal type,
 *     sum up commission and swap from both deals, compute pips and net_pnl
 *  4. Sort results by close_time ascending (chronological order)
 *
 * Idempotent: running this on the same deals multiple times produces the same
 * output, which is required for the upsert-based sync to be safe.
 */
export function normalizeDeals(deals: RawDeal[]): TradeRow[] {
  // Step 1: Group deals by position_id
  // Each entry tracks the opening deal (IN) and closing deal (OUT) separately
  const byPosition = new Map<number, { in: RawDeal | null; out: RawDeal | null }>();

  for (const deal of deals) {
    // Fallback to ticket if position_id is missing (shouldn't happen, defensive)
    const pid = deal.position_id ?? deal.ticket;
    if (!byPosition.has(pid)) {
      byPosition.set(pid, { in: null, out: null });
    }
    const pair = byPosition.get(pid)!;

    if (deal.entry === ENTRY_IN) {
      pair.in = deal; // store the opening deal
    } else if (deal.entry === ENTRY_OUT) {
      // For partial closes: overwrite with the latest OUT deal.
      // This means we capture the final close, not intermediate partials.
      pair.out = deal;
    }
    // entry > 1 exists (e.g. DEAL_ENTRY_INOUT for some brokers) — ignored for now
  }

  // Step 2 & 3: Build TradeRow for each completed position
  const rows: TradeRow[] = [];

  for (const [pid, pair] of Array.from(byPosition.entries())) {
    const outDeal = pair.out;
    const inDeal  = pair.in;

    // Skip positions without an exit deal — still open (or data is incomplete)
    if (!outDeal) continue;

    // DIRECTION LOGIC (MT5 quirk):
    // The exit deal type indicates what action CLOSED the position:
    //   exit type 1 (sell order) → the original trade was a BUY (sold to close)
    //   exit type 0 (buy order)  → the original trade was a SELL (bought to cover)
    const exitType  = outDeal.type ?? 0;
    const direction: "buy" | "sell" = exitType === 1 ? "buy" : "sell";

    // Extract prices and times from the respective deals
    const openPrice  = inDeal?.price  ?? null;
    const closePrice = outDeal.price   ?? null;
    const openTime   = inDeal?.time   ?? null;
    const closeTime  = outDeal.time    ?? null;

    // SL/TP: only available on the entry deal; treat 0 as "not set"
    const sl = (inDeal?.sl && inDeal.sl !== 0) ? inDeal.sl : null;
    const tp = (inDeal?.tp && inDeal.tp !== 0) ? inDeal.tp : null;

    // Duration: only computable if we have both timestamps
    const duration =
      openTime && closeTime ? durationMinutes(openTime, closeTime) : null;

    // FINANCIAL CALCULATIONS:
    // Gross P&L (profit) comes from the exit deal only (MT5 calculates it there)
    const pnl        = outDeal.profit    ?? 0;
    // Commission is split between entry and exit deals — we need both halves
    const commission = (outDeal.commission ?? 0) + (inDeal?.commission ?? 0);
    // Swap is similarly split across both deals
    const swap       = (outDeal.swap      ?? 0) + (inDeal?.swap       ?? 0);
    // Net P&L = what actually lands in the account (commission and swap are negative)
    const netPnl     = pnl + commission + swap;

    // Pips: only calculable if we have both prices
    const pnlPips =
      openPrice !== null && closePrice !== null
        ? calcPips(openPrice, closePrice, direction, outDeal.symbol)
        : null;

    rows.push({
      position_id:      pid,
      ticket:           outDeal.ticket ?? null,
      symbol:           outDeal.symbol,
      direction,
      lot_size:         outDeal.lot_size ?? 0,
      open_price:       openPrice,
      close_price:      closePrice,
      sl,
      tp,
      open_time:        openTime,
      close_time:       closeTime,
      duration_minutes: duration,
      pnl,
      pnl_pips:         pnlPips,
      commission,
      swap,
      net_pnl:          netPnl,
    });
  }

  // Sort by close_time ascending so the array is ready for charts
  rows.sort((a, b) => (a.close_time ?? "").localeCompare(b.close_time ?? ""));
  return rows;
}

/**
 * Converts a TradeRow into the flat object shape expected by Supabase upsert.
 *
 * The account_id is injected here (not in the TradeRow itself) because
 * normalizeDeals doesn't know which account the deals belong to —
 * that context is provided by the sync route.
 *
 * Returns a Record<string, unknown> because Supabase's .upsert() is typed
 * generically. TypeScript will verify column names at the Supabase query level.
 */
export function tradeRowToSupabase(row: TradeRow, accountId: string): Record<string, unknown> {
  return {
    account_id:       accountId,  // FK linking trade to its account
    position_id:      row.position_id,
    ticket:           row.ticket,
    symbol:           row.symbol,
    direction:        row.direction,
    lot_size:         row.lot_size,
    open_price:       row.open_price,
    close_price:      row.close_price,
    sl:               row.sl,
    tp:               row.tp,
    open_time:        row.open_time,
    close_time:       row.close_time,
    duration_minutes: row.duration_minutes,
    pnl:              row.pnl,
    pnl_pips:         row.pnl_pips,
    commission:       row.commission,
    swap:             row.swap,
    net_pnl:          row.net_pnl,
    // Note: user-editable fields (tags, notes, setup_type, mood, mistakes,
    // screenshot_url) are intentionally omitted so syncs don't overwrite
    // the user's journal notes. Supabase upsert leaves existing values alone
    // for columns not included in the upsert payload.
  };
}

File: trade Journal 2.0/tradezella-clone/mac/normalizer.py
"""
mac/normalizer.py — Maps raw MT5 deal data into clean trade rows.

MT5 represents each round-trip trade as TWO deals:
  - Entry deal  (entry == 0, DEAL_ENTRY_IN)  — position opened
  - Exit deal   (entry == 1, DEAL_ENTRY_OUT) — position closed

This module:
  1. Groups all deals by position_id
  2. Pairs each IN deal with its matching OUT deal
  3. Produces one clean TradeRow per completed round trip
  4. Computes derived fields: duration, pnl_pips, net_pnl
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

# ---------------------------------------------------------------------------
# Pip sizes for common symbols
# Override / extend as needed for your instruments
# ---------------------------------------------------------------------------
PIP_SIZES: dict[str, float] = {
    # Forex majors / minors
    "EURUSD": 0.0001, "GBPUSD": 0.0001, "AUDUSD": 0.0001,
    "NZDUSD": 0.0001, "USDCAD": 0.0001, "USDCHF": 0.0001,
    "EURGBP": 0.0001, "EURJPY": 0.01,   "GBPJPY": 0.01,
    "USDJPY": 0.01,   "AUDJPY": 0.01,   "CADJPY": 0.01,
    "CHFJPY": 0.01,   "NZDJPY": 0.01,
    # Metals
    "XAUUSD": 0.1,    "XAGUSD": 0.001,
    # Indices (approximate — adjust per broker)
    "US30":   1.0,    "US500": 0.1,     "NAS100": 0.1,
    "GER40":  0.1,    "UK100": 0.1,
    # Crypto (approximate)
    "BTCUSD": 1.0,    "ETHUSD": 0.1,
}

DEFAULT_PIP_SIZE = 0.0001  # fallback for unknown symbols


def get_pip_size(symbol: str) -> float:
    """Return the pip size for a symbol, stripping broker suffixes like '.r'."""
    clean = symbol.upper().rstrip("M").rstrip(".")
    # Try exact match first, then strip common broker suffixes
    return PIP_SIZES.get(clean, PIP_SIZES.get(symbol.upper(), DEFAULT_PIP_SIZE))


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------

@dataclass
class TradeRow:
    """A single completed round-trip trade ready to insert into Supabase."""
    position_id:      int
    ticket:           int | None       # exit deal ticket
    symbol:           str
    direction:        str              # "buy" | "sell"
    lot_size:         float
    open_price:       float | None
    close_price:      float | None
    sl:               float | None
    tp:               float | None
    open_time:        str | None       # ISO-8601
    close_time:       str | None       # ISO-8601
    duration_minutes: int | None
    pnl:              float            # gross profit from MT5
    pnl_pips:         float | None
    commission:       float
    swap:             float
    net_pnl:          float            # pnl + commission + swap


# ---------------------------------------------------------------------------
# Core normalizer
# ---------------------------------------------------------------------------

ENTRY_IN  = 0   # mt5.DEAL_ENTRY_IN
ENTRY_OUT = 1   # mt5.DEAL_ENTRY_OUT

# MT5 deal type: 0 = buy, 1 = sell
DIRECTION_MAP: dict[int, str] = {0: "buy", 1: "sell"}


def _parse_time(iso: str) -> datetime:
    return datetime.fromisoformat(iso.replace("Z", "+00:00"))


def _duration_minutes(open_iso: str, close_iso: str) -> int:
    delta = _parse_time(close_iso) - _parse_time(open_iso)
    return max(0, int(delta.total_seconds() / 60))


def _pips(open_price: float, close_price: float, direction: str, symbol: str) -> float:
    pip = get_pip_size(symbol)
    raw = (close_price - open_price) / pip
    return round(raw if direction == "buy" else -raw, 1)


def normalize_deals(deals: list[dict[str, Any]]) -> list[TradeRow]:
    """
    Convert a flat list of MT5 deal dicts (from /history) into TradeRows.

    Pairs IN deals with OUT deals by position_id.
    Deals with no matching pair are still included with None for unknown fields
    (e.g. if history only goes back partway and the entry deal is missing).
    """
    # Group by position_id
    by_position: dict[int, dict[str, Any | None]] = {}

    for deal in deals:
        pid = deal.get("position_id") or deal.get("ticket")
        if pid not in by_position:
            by_position[pid] = {"in": None, "out": None}

        entry = deal.get("entry", -1)
        if entry == ENTRY_IN:
            by_position[pid]["in"] = deal
        elif entry == ENTRY_OUT:
            # Keep only the last OUT deal (handles partial closes naively)
            by_position[pid]["out"] = deal

    rows: list[TradeRow] = []

    for pid, pair in by_position.items():
        out_deal = pair["out"]
        in_deal  = pair["in"]

        # Skip positions that never closed (still open — covered by /trades)
        if out_deal is None:
            continue

        # Direction comes from the exit deal type (opposite of entry on a closing deal)
        # MT5 exit deal type: 1 = sell (closing a buy), 0 = buy (closing a sell)
        # So the position direction is the opposite of the exit deal type
        exit_type = out_deal.get("type", 0)
        direction = "buy" if exit_type == 1 else "sell"

        open_price  = in_deal["price"] if in_deal else None
        close_price = out_deal.get("price")
        open_time   = in_deal["time"]  if in_deal else None
        close_time  = out_deal.get("time")

        sl = in_deal.get("sl") if in_deal else None
        tp = in_deal.get("tp") if in_deal else None

        duration = (
            _duration_minutes(open_time, close_time)
            if open_time and close_time
            else None
        )

        pnl        = float(out_deal.get("profit", 0))
        commission = float(out_deal.get("commission", 0)) + (
            float(in_deal.get("commission", 0)) if in_deal else 0.0
        )
        swap = float(out_deal.get("swap", 0)) + (
            float(in_deal.get("swap", 0)) if in_deal else 0.0
        )
        net_pnl = pnl + commission + swap

        pnl_pips = (
            _pips(open_price, close_price, direction, out_deal.get("symbol", ""))
            if open_price is not None and close_price is not None
            else None
        )

        rows.append(
            TradeRow(
                position_id      = int(pid),
                ticket           = out_deal.get("ticket"),
                symbol           = out_deal.get("symbol", ""),
                direction        = direction,
                lot_size         = float(out_deal.get("lot_size", 0)),
                open_price       = open_price,
                close_price      = close_price,
                sl               = sl if sl else None,
                tp               = tp if tp else None,
                open_time        = open_time,
                close_time       = close_time,
                duration_minutes = duration,
                pnl              = pnl,
                pnl_pips         = pnl_pips,
                commission       = commission,
                swap             = swap,
                net_pnl          = net_pnl,
            )
        )

    # Sort by close_time ascending
    rows.sort(key=lambda r: r.close_time or "")
    return rows


def trade_row_to_dict(row: TradeRow, account_id: str) -> dict[str, Any]:
    """Serialize a TradeRow to a dict ready for Supabase upsert."""
    return {
        "account_id":       account_id,
        "position_id":      row.position_id,
        "ticket":           row.ticket,
        "symbol":           row.symbol,
        "direction":        row.direction,
        "lot_size":         row.lot_size,
        "open_price":       row.open_price,
        "close_price":      row.close_price,
        "sl":               row.sl,
        "tp":               row.tp,
        "open_time":        row.open_time,
        "close_time":       row.close_time,
        "duration_minutes": row.duration_minutes,
        "pnl":              row.pnl,
        "pnl_pips":         row.pnl_pips,
        "commission":       row.commission,
        "swap":             row.swap,
        "net_pnl":          row.net_pnl,
    }
